app/javascript/src/Policies/components/CustomPolicy.jsx.orig

```diff
@@ -0,0 +1,107 @@
+// @flow
+
+import * as React from 'react'
+import { useState } from 'react'
+import { UnControlled as CodeMirror } from 'react-codemirror2'
+import SchemaForm from 'react-jsonschema-form'
+import { parsePolicies, fromJson, toJson } from 'Policies/util'
+import type { RegistryPolicy, RawRegistry } from 'Policies/types/Policies'
+import 'codemirror/mode/javascript/javascript'
+import 'Policies/styles/policies.scss'
+
+const cmOptions = {
+  theme: 'default',
+  height: 'auto',
+  viewportMargin: Infinity,
+  mode: {
+    name: 'javascript',
+    json: true,
+    statementIndent: 2
+  },
+  lineNumbers: true,
+  lineWrapping: true,
+  indentWithTabs: false,
+  tabSize: 2
+}
+
+function Editor ({onChange, code}): React.Node {
+  const [ state, setState ] = useState({valid: true, code: toJson(code)})
+
+  const onCodeChange = (editor, metadata, code) => {
+    setState({ valid: true, code })
+    setImmediate(() => {
+      try {
+        onChange(fromJson(code))
+      } catch (err) {
+        setState({ valid: false, code })
+      }
+    })
+  }
+
+  const icon = state.valid ? 'check' : 'times'
+  const cls = state.valid ? 'valid' : 'invalid'
+  const hiddenClass = state.valid ? 'is-hidden' : ''
+  const errorClass = state.valid ? '' : 'CustomPolicy-error'
```

Comment:
Can we use destructuring assignment to check for `state.valid` only once? Or not a good idea due to browser support? Or not too complicated?

```javascript
const [icon, cls, hiddenClass, errorClass] = state.valid ? ['check', 'valid', 'is-hidden', ''] : ['times', 'invalid', '', 'CustomPolicy-error'];
```